### PR TITLE
Add support to ingest log entries to Stackdriver against new "k8s_container" and "k8s_node" resources.

### PR DESCRIPTION
**Pull request by qingling128**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**
Add support to ingest log entries to Stackdriver against new "k8s_container" and "k8s_node" resources.
Also Fluentd 0.14 has some memory leak issues that caused the e2e tests to be flaky. Downgrading to v0.12.

**Special notes for your reviewer**:
We never released any previous version with Fluentd v0.14. Only upgraded it very recently. So this downgrading is not visible to users.

**Release note**:
```release-note
Add support to ingest log entries to Stackdriver against new "k8s_container" and "k8s_node" resources.
```

### CODE REVIEW

**Comment by qingling128** [comment]
Just signed the CLA.

**Comment by qingling128** [comment]
CC: @piosz

**Comment by qingling128** [comment]
Just for the reference, the original PR where we updated to an image using Fluentd v0.14 and adjusted the config file's parser section is at https://github.com/kubernetes/kubernetes/pull/59128.

As part of the downgrading to v0.12, we had to revert the parser section config change.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 1]
```diff
@@ -75,7 +75,19 @@ data:
     <match reform.**>
       @type record_reformer
       enable_ruby true
-      tag raw.kubernetes.${tag_suffix[4].split('-')[0..-2].join('-')}
+      <record>
+        # Extract local_resource_id from tag for 'k8s_container' monitored
+        # resource. The format is:
+        # 'k8s_container.<namespace_name>.<pod_name>.<container_name>'.
+        "logging.googleapis.com/local_resource_id" ${"k8s_container.#{tag_suffix[4].rpartition('.')[0].split('_')[1]}.#{tag_suffix[4].rpartition('.')[0].split('_')[0]}.#{tag_suffix[4].rpartition('.')[0].split('_')[2].rpartition('-')[0]}"}
+        # Rename the field 'log' to a more generic field 'message'. This way the
```
Comment (qingling128) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: For `gke_container` resource, we are sending `log` field as `text_payload` after `time`/`severity`/`stream` are extracted. For every other resources, we are sending `message` field as textPayload if it is the only field.

To assure the same behavior for `k8s_container`, we can either rename the `log` field to `message` in the configuration, or hardcode an exception in the fluent-plugin-google-cloud gem next to the `gke_container` resource.

I'm slightly leaning towards handling it via configuration instead of hardcoding it in the gem. The only issue is that when there are additional fields (other than `time`/`severity`/`stream`/`log`), the behavior might be different. For the `gke_container` resource, we seem to be dropping all other fields (That feel like a bug to me, isn't it?). For the `k8s_container` resource, we will keep the jsonPayload structure and retain the additional fields.

WDYT? @bmoyles0117 @igorpeshansky

Comment (igorpeshansky) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: For `gke_container`, the implementation and the configuration on the Kubernetes side were carefully crafted to work together. This, as you observed, was very fragile.
The semantics you propose is simple and straightforward. We can document that the logging agent expects the text payload in a `message` field, and rename `log` to `message` in the configuration.

File: cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml [hunk 2]
```diff
@@ -37,6 +37,16 @@ spec:
           readOnly: true
         - name: config-volume
           mountPath: /etc/google-fluentd/config.d
+        args:
+        - /bin/sh
+        - -c
+        - STACKDRIVER_METADATA_AGENT_URL="http://${NODE_NAME}:8000" /usr/sbin/google-fluentd
```
Comment (x13n) [inline cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml]: Can't STACKDRIVER_METADATA_AGENT_URL be set via env?

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 3]
```diff
@@ -366,6 +366,11 @@ data:
       </record>
     </filter>
 
+    # Do not collect fluentd's own logs to avoid infinite loops.
+    <match fluent.**>
```
Comment (x13n) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Does that mean no logs from fluentd will be exported?

File: cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml (hunk 2 above)
Comment (x13n) [reply cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml]: Also, double-checking - is this the right port?

**Review by qingling128 — COMMENTED** [review]
All feedback addressed.
* Grouped fluentd config setups into a setup-fluentd method.
* Made metadata_agent_url configurable via env var.
* Added an env var to control whether to ingest logs against old or new resources.

The [contributor guide](https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#6-squashing-and-commit-titles) seems to suggest not to squash unless being suggested by reviewers. So I'll leave these commits separate for now.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap-old.yaml [hunk 4]
```diff
@@ -0,0 +1,425 @@
+# This ConfigMap is used to ingest logs against old resources like
```
Comment (qingling128) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap-old.yaml]: NOTE: This file is roughly a copy of `fluentd-gcp-configmap.yaml`. The only difference is whether to set up `local_resource_id` and enable talking to Metadata Agent. With those, logs are ingested against `k8s_container` and `k8s_node`. Without those, logs are ingested against `gke_container` and `gce_instance`.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 5]
```diff
@@ -51,26 +57,18 @@ data:
       pos_file /var/log/gcp-containers.log.pos
       tag reform.*
       read_from_head true
-      format none
+      format multi_format
```
Comment (qingling128) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: NOTE:  This entire `format` change is to revert the changes we made in https://github.com/kubernetes/kubernetes/pull/59128/files#diff-dec77c261fefaa453d67b0c26b4b07c2 to adapt to Fluentd `v0.14` syntax. As we downgrade back to `v0.12` in this PR, the syntax needs to be reverted as well.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 6]
```diff
@@ -83,7 +81,19 @@ data:
     <match reform.**>
       @type record_reformer
```
Comment (qingling128) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: NOTE: This whole `record_reformer` section sets up the `local_resource_id` for `k8s_container` resource. It's present in `fluentd-gcp-configmap.yaml` only (not `fluentd-gcp-configmap-old.yaml`).

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 7]
```diff
@@ -397,13 +412,19 @@ data:
...
-      labels {
-        # The logging backend will take responsibility for double writing to
-        # the necessary resource types when this label is set.
-        "logging.googleapis.com/k8s_compatibility": "true"
-      }
+      # Use Metadata Agent to get monitored resource.
+      enable_metadata_agent true
     </match>
 
+    # Attach local_resource_id for 'k8s_node' monitored resource.
```
Comment (qingling128) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: NOTE: This entire `record_reformer` section sets up the `local_resource_id` for `k8s_node` resource. It's present in `fluentd-gcp-configmap.yaml` only (not `fluentd-gcp-configmap-old.yaml`).

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 8]
```diff
@@ -397,13 +412,19 @@ data:
       # Use multiple threads for processing.
       num_threads 2
       use_grpc true
-      labels {
-        # The logging backend will take responsibility for double writing to
-        # the necessary resource types when this label is set.
-        "logging.googleapis.com/k8s_compatibility": "true"
-      }
+      # Use Metadata Agent to get monitored resource.
```
Comment (qingling128) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: NOTE: This `enable_metadata_agent true` configuration is present in `fluentd-gcp-configmap.yaml` only (not `fluentd-gcp-configmap-old.yaml`).

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 9]
```diff
@@ -425,14 +446,11 @@ data:
       disable_retry_limit
       num_threads 2
       use_grpc true
-      labels {
-        # The logging backend will take responsibility for double writing to
-        # the necessary resource types when this label is set.
-        "logging.googleapis.com/k8s_compatibility": "true"
-      }
+      # Use Metadata Agent to get monitored resource.
+      enable_metadata_agent true
```
Comment (qingling128) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: NOTE: This `enable_metadata_agent true` configuration is present in `fluentd-gcp-configmap.yaml` only (not `fluentd-gcp-configmap-old.yaml`).

File: cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml [hunk 10]
```diff
@@ -37,6 +37,10 @@ spec:
           readOnly: true
         - name: config-volume
           mountPath: /etc/google-fluentd/config.d
+        args:
```
Comment (qingling128) [inline cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml]: NOTE: `STACKDRIVER_METADATA_AGENT_URL` is set up as an environment variable regardless whether we are ingesting logs via old resources (in which case `STACKDRIVER_METADATA_AGENT_URL` is not needed) or new resources. Just because there seems to be no need to distinguish the case.

Thoughts?

Comment (bmoyles0117) [reply cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml]: I'm ok with setting this unconditionally, however, why aren't we setting this as a normal environment variable? Injecting it at this point is actually problematic as it harms the ability to shut down cleanly.

```
        - env:
          - name: STACKDRIVER_METADATA_AGENT_URL
            value: {{ stackdriver_metadata_agent_url }}
```

Is what I'm suggesting.

File: cluster/gce/gci/configure-helper.sh [hunk 11]
```diff
@@ -2051,6 +2051,33 @@ function setup-coredns-manifest {
...
 
+# Sets up the manifests of Fluentd configmap and yamls for k8s addons.
+function setup-fluentd {
+  local -r dst_dir="$1"
+  local -r fluentd_gcp_yaml="${dst_dir}/fluentd-gcp/fluentd-gcp-ds.yaml"
+  # Ingest logs against new resources like "k8s_container" and "k8s_node" if
+  # LOGGING_STACKDRIVER_RESOURCES is "new".
+  # Ingest logs against old resources like "gke_container" and "gce_instance" if
+  # LOGGING_STACKDRIVER_RESOURCES is "old".
+  if [[ "${LOGGING_STACKDRIVER_RESOURCES:-}" == "new" ]]; then
```
Comment (bmoyles0117) [inline cluster/gce/gci/configure-helper.sh]: I confused RESOURCES with actual resources as defined by Kubernetes, this is a nitpick on my part, but I'd petition to change this to `LOGGING_STACKDRIVER_RESOURCE_TYPES`, or anything else that doesn't blur the line of cpu / memory resources. Even resource types might be confusing. This could all just be unnecessary noise, ignore me if that keeps things rolling :)

https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap-old.yaml (hunk 4 above)
Comment (bmoyles0117) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap-old.yaml]: +1 to what you've done here.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 12]
```diff
@@ -1,3 +1,9 @@
+# This ConfigMap is used to ingest logs against new resources like
+# "k8s_container" and "k8s_node".
+# See the ConfigMap for ingesting logs against old resources like
+# "gke_container" and "gce_instance" in fluentd-gcp-configmap-old.yaml.
+# Use $LOGGING_STACKDRIVER_RESOURCES to control which ConfigMap will be
```
Comment (bmoyles0117) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: I would expand on specifically setting the value to "new" or "old".

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 1 above)
Comment (bmoyles0117) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: +1 to what Igor said.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 3 above)
Comment (bmoyles0117) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Ling, have you confirmed that this section does what it tries to do? As far as I'm aware, fluent logs are emitted to stdout/stderr, making this tag meaningless. If you've confirmed this behavior works as intended great, but if it doesn't add any value, I would suggest removing it.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 13]
```diff
@@ -397,13 +412,19 @@ data:
...
+      # Use Metadata Agent to get monitored resource.
+      enable_metadata_agent true
     </match>
 
+    # Attach local_resource_id for 'k8s_node' monitored resource.
+    <filter **>
+      @type record_transformer
+      enable_ruby true
+      <record>
+        "logging.googleapis.com/local_resource_id" ${"k8s_node.#{ENV['NODE_NAME']}"}
```
Comment (bmoyles0117) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Careful, I don't think `NODE_NAME` exists, it would need to be added to the daemonset yaml as an environment variable pulled from the downwards api. This is what I've used historically.

```
        - env:
          - name: NODE_NAME
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: spec.nodeName
```

**Review by bmoyles0117 — CHANGES_REQUESTED** [review]

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 3 above)
Comment (qingling128) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Yes, this indeed takes effect. I encountered infinite loops and this change fixed it.

A sample ingested fluentd log looks like below:

```
{
 insertId:  "9ppb4lg2hn1rtn"  
 logName:  "projects/stackdriver-kubernetes-1337/logs/fluent.info"  
 receiveTimestamp:  "2018-04-04T01:27:05.638928314Z"  
 resource: {
  labels: {
   cluster_name:  "stackdriver-metadata-e2e"    
   location:  "us-central1-a"    
   node_name:  "gke-stackdriver-metadata-default-pool-c622577e-vlmk"    
   project_id:  "stackdriver-kubernetes-1337"    
  }
  type:  "k8s_node"   
 }
 textPayload:  "following tail of /var/log/containers/metadata-agent-api-availability-test-p469t_metadata-agent-api-availability-test_metadata-agent-api-availability-test-ef878dbc383e0e59b558103f2acfa446c4ca8d8520695460f5764e5a5719b468.log"  
 timestamp:  "2018-04-04T01:27:04Z"  
}
```

As we can see, Fluentd logs are tagged as `fluent.info`, `fluent.error` and `fluent.warn`. 

This is because Fluentd logs are [emitted as Fluentd events directly](https://github.com/fluent/fluentd/blob/b503fdbeb670ca71cbc5de2fea43702da55b10db/lib/fluent/log.rb#L488).

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 13 above)
Comment (qingling128) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Oops. This used to be there. I removed it in the last commit thinking it was only used by the `metadata_agent_url` and no longer needed. Just added it back.

File: cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml (hunk 10 above)
Comment (qingling128) [reply cluster/addons/fluentd-gcp/fluentd-gcp-ds.yaml]: Good point. Changed.

File: cluster/gce/gci/configure-helper.sh (hunk 11 above)
Comment (qingling128) [reply cluster/gce/gci/configure-helper.sh]: That's a valid point. Renamed to `LOGGING_STACKDRIVER_RESOURCE_TYPES`. It's not ideal, but at least less confusing.

**Review by qingling128 — COMMENTED** [review]
All feedback addressed. PTAL.

**Comment by qingling128** [comment]
@x13n - Some questions regarding to the e2e test coverage:
Right now we are ingesting logs against old resources (`gke_container` and `gce_instance`) when `$LOGGING_STACKDRIVER_RESOURCE_TYPES` is not `new`, which is the default behavior.
Ideally e2e should test both cases ("new" and "old"). My hunch is that it's only testing one case ("old") today and some additional work might be needed there. If so, is the work trivial to cover both cases? Is that something we must worry about for the patch release?

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap-old.yaml [hunk 14]
```diff
@@ -0,0 +1,425 @@
...
+      buffer_chunk_limit 1M
+      buffer_queue_limit 2
+      flush_interval 5s
+      max_retry_wait 30
+      disable_retry_limit
+      num_threads 2
+      use_grpc true
+    </match>
+metadata:
+  name: fluentd-gcp-config-v1.2.5
```
Comment (x13n) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap-old.yaml]: Can these configmaps be named differently? If the name stays the same, it will be harder to inspect the objects - we should be able to tell which one is used, just by looking at the name. Trying to infer this from configmap contents will be tedious. How about fluentd-gcp-legacy-config? or ...-old-config?

Comment (qingling128) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap-old.yaml]: Good point. Changed.

**Review by qingling128 — COMMENTED** [review]
PTAL.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap-old.yaml [hunk 15]
```diff
@@ -0,0 +1,425 @@
...
+      buffer_chunk_limit 1M
+      buffer_queue_limit 2
+      flush_interval 5s
+      max_retry_wait 30
+      disable_retry_limit
+      num_threads 2
+      use_grpc true
+    </match>
+metadata:
+  name: fluentd-gcp-old-config-v1.2.5
```
Comment (bmoyles0117) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap-old.yaml]: Small nit, `s/fluentd-gcp-old-config/fluentd-gcp-config-old/` to be consistent with the naming of the files.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml [hunk 16]
```diff
@@ -83,7 +81,19 @@ data:
...
+        # resource. The format is:
+        # 'k8s_container.<namespace_name>.<pod_name>.<container_name>'.
+        "logging.googleapis.com/local_resource_id" ${"k8s_container.#{tag_suffix[4].rpartition('.')[0].split('_')[1]}.#{tag_suffix[4].rpartition('.')[0].split('_')[0]}.#{tag_suffix[4].rpartition('.')[0].split('_')[2].rpartition('-')[0]}"}
+        # Rename the field 'log' to a more generic field 'message'. This way the
+        # fluent-plugin-google-cloud knows to flatten the field as textPayload
+        # instead of jsonPayload after extracting 'time', 'severity' and
+        # 'stream' from the record.
+        message ${record['log']}
+      </record>
+      tag ${if record['stream'] == 'stderr' then 'raw.kubernetes.stderr' else 'raw.kubernetes.stdout' end}
```
Comment (bmoyles0117) [inline cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Let's open a quick discussion in person today to confirm with the logging backend that we shouldn't use the container name as the tag. Container name is consistent across pods so it wouldn't flood the UI dropdown, and it might be a "better" initial user experience. If we can't get an answer in ~30 minutes I'm fine with proceeding as-is.

File: cluster/gce/gci/configure-helper.sh [hunk 17]
```diff
@@ -2051,6 +2051,36 @@ function setup-coredns-manifest {
...
+    local -r fluentd_gcp_configmap_yaml="${dst_dir}/fluentd-gcp/fluentd-gcp-configmap.yaml"
+    fluentd_gcp_configmap_name="fluentd-gcp-config"
+  else
+    local -r fluentd_gcp_configmap_yaml="${dst_dir}/fluentd-gcp/fluentd-gcp-configmap-old.yaml"
+    fluentd_gcp_configmap_name="fluentd-gcp-old-config"
+  fi
+  sed -i -e "s@{{ fluentd_gcp_configmap_name }}@${fluentd_gcp_configmap_name}@g" "${fluentd_gcp_yaml}"
+  fluentd_gcp_version="${FLUENTD_GCP_VERSION:-0.2-1.5.30-1-k8s}"
+  sed -i -e "s@{{ fluentd_gcp_version }}@${fluentd_gcp_version}@g" "${fluentd_gcp_yaml}"
+  if [[ "${STACKDRIVER_METADATA_AGENT_URL:-}" != "" ]]; then
```
Comment (bmoyles0117) [inline cluster/gce/gci/configure-helper.sh]: I think this needs to be added to `PROVIDER_VARS` as well, @x13n @kawych to confirm.

**Review by bmoyles0117 — CHANGES_REQUESTED** [review]

File: cluster/gce/gci/configure-helper.sh (hunk 17 above)
Comment (x13n) [reply cluster/gce/gci/configure-helper.sh]: Sorry, I missed this one. Yes, if this can change, then it should be added there.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap-old.yaml (hunk 15 above)
Comment (qingling128) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap-old.yaml]: Changed.

File: cluster/gce/gci/configure-helper.sh (hunk 17 above)
Comment (qingling128) [reply cluster/gce/gci/configure-helper.sh]: Added.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 16 above)
Comment (qingling128) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: The requirement to not use the container name as the tag initially came from the Logging Backend folks. Their standpoint probably won't change from the technical perspective. If we do decide to change, it's probably driven by UX experience need. Maybe we should talk to PM?

**Review by qingling128 — COMMENTED** [review]
PTAL.

File: cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml (hunk 16 above)
Comment (qingling128) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: On a second thought, even if we want to fulfill the UX need, the container name based dropped down should be reflected in the UI dropdown `Kubernetes Container => Location => Cluster Name => ...` instead of the logname dropdown. We don't have to include it in the logname (backend will have to index the same thing twice).

Comment (qingling128) [reply cluster/addons/fluentd-gcp/fluentd-gcp-configmap.yaml]: Per discussion offline with @bmoyles0117 , we will not include container names in the tags. However, the tags should be changed to `stderr` and `stdout` instead of `kubernetes.stderr` and `kubernetes.stdout`.

The change is in. PTAL.

**Review by qingling128 — COMMENTED** [review]
Pushed another commit to change container log tags from `kubernetes.stderr` and `kubernetes.stdout` to `stderr` and `stdout`.

Does it look better now? PTAL. @x13n @bmoyles0117

**Review by bmoyles0117 — APPROVED** [review]
LGTM

**Comment by x13n** [comment]
Regarding tests - yes, we need to test both new and old resources, but this will be a separate PR to test-infra (namely, new entry in https://github.com/kubernetes/test-infra/blob/master/jobs/config.json#L337).

/lgtm

**Comment by x13n** [comment]
@MaciekPytel Can you approve for cluster/gce part?

**Comment by MaciekPytel** [comment]
@qingling128 Please update release note, so it's clear this is specific to Stackdriver and not some general change in k8s logging (current version is way too specific for the target audience of Kubernetes users).

/approve

**Comment by qingling128** [comment]
@MaciekPytel - Changed the PR title and release note. Should I squash everything into one commit now?

**Comment by MaciekPytel** [comment]
Oh yes, I completely missed the number of commits. Please squash them.

/hold 
So it doesn't merge before squashing.

**Comment by kawych** [comment]
/lgtm
@x13n is out of office, so approving instead of him as he lgtmed before squashing commits.

**Comment by qingling128** [comment]
Thanks @kawych!

@MaciekPytel - PTAL.

**Comment by qingling128** [comment]
Finally made it into the queue (there was some flakiness of the test) and merged. A cherrypick PR has been created as well.

**Comment by StevenACoffman** [comment]
@qingling128 Thank you for your work on this. What is the issue tracking the fluentd memory leak so we can know when to try to upgrade again?

**Comment by StevenACoffman** [comment]
@monotek are you aware of any issue tracking the fluentd memory leak referred to here?

**Comment by qingling128** [comment]
@StevenACoffman @monotek - I've created an internal ticket to track it. Will report it with Fluentd as well and link it here. (Meant to do it earlier, but got distracted by the cherrypick issues)

**Comment by qingling128** [comment]
Created https://github.com/fluent/fluentd/issues/1941 to track the memory leak.

**Comment by StevenACoffman** [comment]
Hmmm... I am not seeing that with latest (only running for a few days) and there have been several leak fixes in versions in between:
* [see ChangeLog|https://github.com/fluent/fluentd/blob/master/CHANGELOG.md]

**Comment by qingling128** [comment]
@StevenACoffman - That sounds promising! We are also setting up some soak test for various Fluentd versions including the latest (v1.1). Some additional testing work is needed from our side to bump a major Fluentd version. We might not have resources for it this quarter, but it's definitely on our roadmap.
